### DISCUSSION TITLE
M2M interface needs to be more ergonomic

### DISCUSSION THREAD

**Discussion by colemahatmccreary** [directus/directus#13990] [Feature Requests]
TLDR: Users should be prompted for all 'junction table' data directly from the 'Select Existing Records' interface. 

Let's say I have two models, 'Shipment' and 'Products', each with many fields of their own. 

I also have a 'Shipment Products' model, creating an M2M relationship between products and shipments (i.e. "which items go in this shipment?").

This 'Shipment Products' junction table includes an FK for each of the main tables, plus additional fields indicating things like the **quantity** of the item to be shipped, the **MSRP** of the item at the time it was included, and so on. Not every relationship requires add'l data but many will.

Generating these connections via the 'Add Existing' button on the 'Shipment' record provides a less than ideal flow for accomplishing this:

1. Directus displays a list of existing 'Product' records, user selects (via checkbox) those to which you want to create a relationship (image 1)
2. Once selected, confirm and return to the Shipment record. You'll now see the selected products staged in the 'Shipment' record's M2M interface (image 2 below). _This is where it gets clunky_.
3. To complete the 'Shipment Product' records (i.e. add the quantities + any other data required), you then need to click each of these staged products one by one to display the 'edit' view, where you need to scroll past every single 'Product' field to reach the bottom, where any fields from the 'Shipment Products' table will be available (without any real indication that's what they are).

Seems like the best way to streamline this would be to check if the junction table has any add'l fields beyond the FK fields as soon as the checkbox is clicked in Step 1. If so, a modal should pop then and there, providing each of these fields for you to populate. That way, by the time you leave the list view from Step 1 all of the relevant data has been collected. At the very minimum, you could put the junction table fields at the very top of the editor rather than the bottom.

Image 1:
<img width="850" alt="Screen Shot 2022-06-19 at 10 33 19 PM" src="https://user-images.githubusercontent.com/40645344/174531805-5cf65391-d8e7-4556-a89c-73c902699499.png">

Image 2: 
<img width="592" alt="Screen Shot 2022-06-19 at 10 31 58 PM" src="https://user-images.githubusercontent.com/40645344/174532697-03143bd8-0e7c-4ed8-9fdf-233616db7953.png">

**Comment by benhaynes** [comment]
I agree that the flow you describe would be an improvement... I wonder if this should be an option on the relational interface (whether to force a prompt for the junction fields after selection). 🤔

**Reply by colemahatmccreary** [reply]
Sure, or it could just be inferred from the schema like other things in Directus. Not sure of the drawbacks there. 

Either way I think this should be default behavior when add’l fields do exist, since you’re essentially just creating new records.

**Reply by benhaynes** [reply]
Yeah, just hard to understand context from a schema... is it extra metadata (eg: notes), or required data (eg: quantity). Intelligent defaults are always a great baseline... but adding the ability to customize can be huge, at times.

**Reply by colemahatmccreary** [reply]
Right, right that makes sense. Sounds like an option would be best!

**Comment by paulboudewijn** [comment]
> At the very minimum, you could put the junction table fields at the very top of the editor rather than the bottom.

I think this is a fairly simple change, which makes M2M interface way more useable.

**Reply by colemahatmccreary** [reply]
Youre in luck - a config option was just added to the m2m field, allowing you to select whether junction fields show on top or bottom of the editor.

**Comment by colemahatmccreary** [comment]
They actually just added a config option to the m2m, allowing you to select
whether the junction fields show on top or bottom of the editor

On Wed, Sep 21, 2022 at 2:30 AM Paul Boudewijn ***@***.***>
wrote:

> At the very minimum, you could put the junction table fields at the very
> top of the editor rather than the bottom.
>
> I think this is a fairly simple change, which makes M2M interface way more
> useable.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/directus/directus/discussions/13990#discussioncomment-3698325>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJWDFYFBDIJYFE4VR3JZVGDV7LIT3ANCNFSM5ZHVG3MQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Cole Mahat McCreary
c: +14145880424
e: ***@***.***

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
